apis/projectcontour/v1alpha1/contourconfig_helpers.go

```diff
@@ -106,22 +104,51 @@ func (e *EnvoyConfig) Validate() error {
 		}
 	}
 
-	// CipherSuites
+	// Envoy TLS configuration
 	if e.Listener != nil && e.Listener.TLS != nil {
-		var invalidCipherSuites []string
-		for _, c := range e.Listener.TLS.CipherSuites {
-			if _, ok := config.ValidTLSCiphers[string(c)]; !ok {
-				invalidCipherSuites = append(invalidCipherSuites, string(c))
-			}
-		}
-		if len(invalidCipherSuites) > 0 {
-			return fmt.Errorf("invalid cipher suites %q", invalidCipherSuites)
-		}
+		return e.Listener.TLS.Validate()
+	}
+
+	return nil
+}
+
+// Validate ensures EnvoyTLS configuration is valid.
+func (e *EnvoyTLS) Validate() error {
+	if e.MinimumProtocolVersion != "" && e.MinimumProtocolVersion != "1.2" && e.MinimumProtocolVersion != "1.3" {
+		return fmt.Errorf("invalid TLS minimum protocol version %q", e.MinimumProtocolVersion)
 	}
 
+	var invalidCipherSuites []string
+	for _, c := range e.CipherSuites {
+		if _, ok := ValidTLSCiphers[c]; !ok {
+			invalidCipherSuites = append(invalidCipherSuites, c)
+		}
+	}
+	if len(invalidCipherSuites) > 0 {
+		return fmt.Errorf("invalid cipher suites %q", invalidCipherSuites)
+	}
 	return nil
 }
 
+// SanitizedCipherSuites returns a deduplicated list of TLS ciphers.
+func (e *EnvoyTLS) SanitizedCipherSuites() []string {
+	if len(e.CipherSuites) == 0 {
+		return DefaultTLSCiphers
+	}
+
+	// Not using the set abstraction here so we can maintain the
+	// list order.
+	uniqueCiphers := map[string]struct{}{}
```

Comment:
Guessing the code changed after the comment was added here, since you're maintaining the ordered list in a slice -- can delete the comment, and optionally actually use a sets.String if you want (slightly nicer API but either way's fine).

Reply:
ah, this is how it originally was in the pkg/config version and I tried using the sets.String api but that gives you back a sorted list of set elements (or "unsorted"/how they come out of the internal map), so the comment was a bit of a warning to not try to use the sets package

Reply:
Yeah I just meant you could replace the usage of the map with a sets.String, but agree you'll need to keep the slice too in order to preserve order.

Reply:
I think it's fine as-is, if anyone has to touch this they can figure it out.

Reply:
ah i see what you mean now